### PR TITLE
gh-117376: Make `Py_DECREF` a macro in ceval.c in free-threaded build

### PR DESCRIPTION
**Pull request by colesbury**
`Py_DECREF` and `PyStackRef_CLOSE` are now implemented as macros in the free-threaded build in ceval.c. There are two motivations;

 * MSVC has problems inlining functions in ceval.c in the PGO build.

 * We will want to mark escaping calls in order to spill the stack pointer in ceval.c and we will want to do this around `_Py_Dealloc` (or `_Py_MergeZeroLocalRefcount` or `_Py_DecRefShared`), not around the entire `Py_DECREF` or `PyStackRef_CLOSE` call.

* Issue: gh-117376

### CODE REVIEW

**Review by DinoV — APPROVED** [review]
LGTM!

**Review by Fidget-Spinner — APPROVED** [review]

**Comment by neonene** [comment]
Can this PR forceinline `Py_NewRef()`, or switch back to `Py_INCREF()` in `pycore_stackref.h`?  MSVC fails to inline `Py_NewRef()` in `ceval.c`, which seems to have some impact.  cc @mdboom

**Comment by mdboom** [comment]
I'm going to run this on the benchmarking infrastructure to confirm it helps.  (This will take about 2 hours to get results, and I'll report back).

> Can this PR forceinline `Py_NewRef()`, or switch back to `Py_INCREF()` in `pycore_stackref.h`? MSVC fails to inline `Py_INCREF` wrappers in `ceval.c`, which seems to have some impact. cc @mdboom
> 
> EDIT: This may not be significant before restoring [gh-121338](https://github.com/python/cpython/issues/121338) or applying `/Os` to `pythoncore.vcxproj` at `PGInstrument` phase.

I agree we should fix this, once we confirm it still matters after #121338 was applied.  I think that can be applied as a separate issue from this one, though, and doesn't have to hold this one up.

File: Include/internal/pycore_stackref.h [hunk 1]
```diff
@@ -191,18 +191,15 @@ PyStackRef_FromPyObjectImmortal(PyObject *obj)
...
-    }
-    Py_DECREF(PyStackRef_AsPyObjectBorrow(stackref));
-}
+#   define PyStackRef_CLOSE(stackref) \
+        do { \
+            _PyStackRef _close_tmp = (stackref); \
+            if (!PyStackRef_IsDeferred(_close_tmp)) { \
+                Py_DECREF(PyStackRef_AsPyObjectBorrow(_close_tmp)); \
+            } \
+        } while (0)
```
Comment (picnixz) [inline Include/internal/pycore_stackref.h]: PEP 7 requires to:

- use uppercase letters for macro parameters,
- align line continuations `\`

**Review by picnixz — COMMENTED** [review]
Just some comment on using PEP 7, although it's not mandatory for old code to follow it but since you're anyway editing things... Line continuations being aligned could be ignored, but having macro parameters names in upper case is usually helpful when you read the macro definition.

**Comment by mdboom** [comment]
The [results are here](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20240813-3.14.0a0-4b27f3a-NOGIL).  tl;dr: This definitely makes things moderately faster (around 1%) on the pyperformance suite on free-threaded builds.  So +1 from me.  Note there are no mac results because our machine is currently down, and no Windows results because PGO is broken for free-threaded builds on Windows.

**Review by mdboom — APPROVED** [review]

**Pull request merged**